terraform/transform_orphan_count.go

```diff
@@ -50,20 +50,47 @@ func (t *OrphanResourceCountTransformer) Transform(g *Graph) error {
 }
 
 func (t *OrphanResourceCountTransformer) transformForEach(haveKeys map[addrs.InstanceKey]struct{}, g *Graph) error {
+	// If there is a no-key node, add this to the graph first,
+	// so that we can create edges to it in subsequent (StringKey) nodes.
+	// This is because the last item determines the resource mode for the whole resource,
+	// so if this (non-deterministically) happens to end up as the last one,
+	// that will change the resource's EachMode and our addressing for our instances
+	// will not work as expected
+	_, hasNoKeyNode := haveKeys[addrs.NoKey]
+	var noKeyNode dag.Vertex
+	if hasNoKeyNode {
+		abstract := NewNodeAbstractResourceInstance(t.Addr.Instance(addrs.NoKey))
+		noKeyNode = abstract
+		if f := t.Concrete; f != nil {
+			noKeyNode = f(abstract)
+		}
+		g.Add(noKeyNode)
+	}
+
 	for key := range haveKeys {
 		s, _ := key.(addrs.StringKey)
 		// If the key is present in our current for_each, carry on
 		if _, ok := t.ForEach[string(s)]; ok {
 			continue
 		}
 
+		// If the key is no-key, we have already added it, so skip
```

Comment:
I would put this check right at the start of the loop. I think it looks better at least to indicate it has no bearing on the other other `ForEach` check and we know we've already added it.

Reply:
I agree, updating